src/lib/specs-crawler.js

```diff
@@ -294,7 +294,7 @@ async function saveSpecResults(spec, settings) {
 async function crawlList(speclist, crawlOptions) {
     // Make a shallow copy of crawl options object since we're going
     // to modify properties in place
-    crawlOptions = Object.assign({}, crawlOptions);
+    crawlOptions = Object.assign({speclist}, crawlOptions);
```

Comment:
We've never been very clean on separating inputs and outputs. Unless I missed a copy somewhere, we directly store the results of the crawl within the spec objects in `speclist`, at least temporarily while the results get saved to disk. This means `speclist` may also contain a few expanded crawl results (or all of them when `useCrawl` is set or when the user does not specify an output folder), up to >50MB of data.

Here, we're asking Puppeteer to send the list to a separate Chrome process, so Puppeteer will need to serialize/deserialize the list each time. That works but that seems like a waste of processing time and memory.

I think we should rather make a deeper copy here and create a new list of spec objects that only contain core metadata about each spec (and that won't be touched by the crawl). That will still send >500KB of data but that should be fine.

I can live with it, but it could perhaps be argued that the extraction code should remain atomic in the sense that its results should only depend on information about the spec and on the spec's contents (extraction code is not the easiest bit to debug since it runs in a headless Chrome instance, so better keep the logic simple). An alternative would be to split the extraction of links into two parts:
1. The extraction of the raw links in Chrome through `extract-links.mjs` without attempting to link them back to specs.
2. A post-processing module that associates extracted links with specs and runs in the Node.js process. Post-processing is a good place to run logic that needs to look at other specs.

Reply:
That approach makes more sense indeed; I've updated the PR to that effect